lxd/db/warnings.go

```diff
@@ -0,0 +1,395 @@
+//go:build linux && cgo && !agent
+// +build linux,cgo,!agent
+
+package db
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/pborman/uuid"
+	"github.com/pkg/errors"
+
+	"github.com/lxc/lxd/lxd/db/cluster"
+	"github.com/lxc/lxd/lxd/db/query"
+	"github.com/lxc/lxd/shared"
+)
+
+// Warning is a value object holding db-related details about a warning.
+type Warning struct {
+	ID             int
+	Node           string
+	Project        string
+	EntityTypeCode int
+	EntityID       int
+	UUID           string
+	TypeCode       int
+	Status         int
+	FirstSeenDate  time.Time
+	LastSeenDate   time.Time
+	LastMessage    string
+	Count          int
+}
+
+var warningCreate = cluster.RegisterStmt(`
+INSERT INTO warnings (node_id, project_id, entity_type_code, entity_id, uuid, type_code, status, first_seen_date, last_seen_date, last_message, count)
+  VALUES ((SELECT nodes.id FROM nodes WHERE nodes.name = ?), (SELECT projects.id FROM projects WHERE projects.name = ?), ?, ?, ?, ?, ?, ?, ?, ?, ?)
+`)
+
+var warningDelete = cluster.RegisterStmt(`
+DELETE FROM warnings WHERE uuid = ?
+`)
+
+var warningDeleteByStatus = cluster.RegisterStmt(`
+DELETE FROM warnings WHERE status = ?
+`)
+
+var warningID = cluster.RegisterStmt(`
+SELECT warnings.id FROM warnings LEFT JOIN nodes ON warnings.node_id = nodes.id LEFT JOIN projects ON warnings.project_id = projects.id
+  WHERE warnings.uuid = ?
+`)
+
+// CreateOrUpdateWarning creates or updates a warning.
+func (c *Cluster) CreateOrUpdateWarning(nodeName string, projectName string, entityTypeCode int, entityID int, typeCode WarningType, message string) error {
```

Comment:
I've seen the use of "upsert" in the code base for this type of "create or update" action, perhaps "UpsertWarning" would be more consistent with that?

Reply:
`SetWarning` seems wrong as we are often creating warnings. But `upsert` might be a good choice.